spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/eventhub/AbstractEventHubTemplate.java

```diff
@@ -0,0 +1,111 @@
+    private final EventHubClientFactory clientFactory;
+
+    @Getter
+    @Setter
+    private EventHubMessageConverter messageConverter = new EventHubMessageConverter();
+
+    @Getter
+    @Setter
+    private StartPosition startPosition = StartPosition.LATEST;
+
+    @Setter
+    @Getter
+    private CheckpointMode checkpointMode = CheckpointMode.BATCH;
+
+    AbstractEventHubTemplate(EventHubClientFactory clientFactory) {
+        this.clientFactory = clientFactory;
+    }
+
+    private static EventProcessorOptions buildEventProcessorOptions(StartPosition startPosition) {
+        EventProcessorOptions options = EventProcessorOptions.getDefaultOptions();
+
+        if (startPosition == StartPosition.EARLISET) {
+            options.setInitialPositionProvider((s) -> EventPosition.fromStartOfStream());
+        } else /* StartPosition.LATEST */ {
+            options.setInitialPositionProvider((s) -> EventPosition.fromEndOfStream());
+        }
+
+        return options;
+    }
+
+    public <T> CompletableFuture<Void> sendAsync(String eventHubName, @NonNull Message<T> message,
+            PartitionSupplier partitionSupplier) {
+        Assert.hasText(eventHubName, "eventHubName can't be null or empty");
+        EventData eventData = messageConverter.fromMessage(message, EventData.class);
+        try {
+            EventHubClient client = this.clientFactory.getEventHubClientCreator().apply(eventHubName);
+
+            if (partitionSupplier == null) {
+                return client.send(eventData);
+            } else if (!Strings.isNullOrEmpty(partitionSupplier.getPartitionId())) {
+                return this.clientFactory.getPartitionSenderCreator()
+                                         .apply(Tuple.of(client, partitionSupplier.getPartitionId())).send(eventData);
+            } else if (!Strings.isNullOrEmpty(partitionSupplier.getPartitionKey())) {
+                return client.send(eventData, partitionSupplier.getPartitionKey());
+            } else {
+                return client.send(eventData);
+            }
+        } catch (EventHubRuntimeException e) {
+            LOGGER.error(String.format("Failed to send to '%s' ", eventHubName), e);
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+        }
+    }
+
+    protected void register(Tuple<String, String> nameAndGroup, EventHubProcessor eventProcessor) {
+        EventProcessorHost host = this.clientFactory.getProcessorHostCreator().apply(nameAndGroup);
+        host.registerEventProcessorFactory(context -> eventProcessor, buildEventProcessorOptions(startPosition));
+    }
+
+    protected void unregister(Tuple<String, String> nameAndGroup) {
+        this.clientFactory.getProcessorHostCreator().apply(nameAndGroup).unregisterEventProcessor()
+                          .whenComplete((s, t) -> {
+                              if (t != null) {
```

Comment:
seems this piece of code pattern appears more than one place, wonder if we can add one `Function` for this.

Reply:
only log might not need another function